sdk/shared/src/main/scala/org/ergoplatform/sdk/ContractTemplate.scala

```diff
@@ -0,0 +1,476 @@
+      val expressionBytes = expressionTreeWriter.toBytes
+      w.putUInt(expressionBytes.length)
+      w.putBytes(expressionBytes)
+    }
+
+    override def parse(r: SigmaByteReader): ContractTemplate = {
+      val treeVersion = r.getOption(r.getUByte().toByte)
+      val name = parseString(r)
+      val description = parseString(r)
+
+      val nConstants = r.getUInt().toInt
+      val constTypes: IndexedSeq[SType] = {
+        if (nConstants > 0) {
+          val res = safeNewArray[SType](nConstants)
+          cfor(0)(_ < nConstants, _ + 1) { i =>
+            res(i) = TypeSerializer.deserialize(r)
+          }
+          res
+        } else {
+          SType.EmptySeq
+        }
+      }
+      val constValues: Option[IndexedSeq[Option[SType#WrappedType]]] = r.getOption((() => {
+        if (nConstants > 0) {
+          val res = safeNewArray[Option[SType#WrappedType]](nConstants)
+          cfor(0)(_ < nConstants, _ + 1) { i =>
+            res(i) = r.getOption((() => DataSerializer.deserialize(constTypes(i), r))())
+          }
+          res
+        } else {
+          Array.empty[Option[SType#WrappedType]]
+        }
+      })())
+
+      val nParameters = r.getUInt().toInt
+      val parameters: IndexedSeq[Parameter] = {
+        if (nParameters > 0) {
+          val res = safeNewArray[Parameter](nParameters)
+          cfor(0)(_ < nParameters, _ + 1) { i =>
+            res(i) = Parameter.serializer.parse(r)
+          }
+          res
+        } else {
+          Parameter.EmptySeq
+        }
+      }
+
+      // Populate constants in constantStore so that the expressionTree can be deserialized.
+      val constants = constTypes.indices.map(i => {
+        val t = constTypes(i)
+        DeserializationSigmaBuilder.mkConstant(defaultOf(t), t)
+      })
+      constants.foreach(c => r.constantStore.put(c)(DeserializationSigmaBuilder))
+
+      val _ = r.getUInt().toInt
+      val expressionTree = ValueSerializer.deserialize(r)
+      CheckDeserializedScriptIsSigmaProp(expressionTree)
+
+      ContractTemplate(
+        treeVersion, name, description,
+        constTypes, constValues, parameters,
+        expressionTree.toSigmaProp)
+    }
+  }
+
+  object jsonEncoder extends JsonCodecs {
+
+    implicit val sTypeEncoder: Encoder[SType] = Encoder.instance({ tpe =>
+      Json.fromString(tpe.toTermString)
+    })
+
+    implicit val sTypeDecoder: Decoder[SType] = Decoder.instance({ implicit cursor =>
+      fromTry(Try.apply(SigmaParser.parseType(cursor.value.asString.get)))
+    })
+
+    implicit val encoder: Encoder[ContractTemplate] = Encoder.instance({ ct =>
+      val expressionTreeWriter = SigmaSerializer.startWriter()
+      ValueSerializer.serialize(ct.expressionTree, expressionTreeWriter)
+
+      Json.obj(
+        "treeVersion" -> ct.treeVersion.asJson,
+        "name" -> Json.fromString(ct.name),
+        "description" -> Json.fromString(ct.description),
+        "constTypes" -> ct.constTypes.asJson,
```

Comment:
If sTypeEncoder is used here, than it is wrong. In EIP5 the string should be hex coming from TypeSerializer.

Reply:
Done, updated to use TypeSerializer.